beater/v2_handler.go

```diff
@@ -0,0 +1,269 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/elastic/apm-server/utility"
+
+	"github.com/elastic/apm-server/transform"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/validation"
+	"github.com/santhosh-tekuri/jsonschema"
+
+	"github.com/elastic/apm-server/decoder"
+	er "github.com/elastic/apm-server/model/error"
+	"github.com/elastic/apm-server/model/metadata"
+	"github.com/elastic/apm-server/model/metric"
+	"github.com/elastic/apm-server/model/span"
+	"github.com/elastic/apm-server/model/transaction"
+)
+
+const batchSize = 10
+
+func (v v2Route) Handler(beaterConfig *Config, report reporter) http.Handler {
+	reqDecoder := v.configurableDecoder(
+		beaterConfig,
+		func(*http.Request) (map[string]interface{}, error) { return map[string]interface{}{}, nil },
+	)
+
+	v2Handler := v2Handler{
+		requestDecoder: reqDecoder,
+		tconfig:        v.transformConfig(beaterConfig),
+	}
+
+	return v.wrappingHandler(beaterConfig, v2Handler.Handle(beaterConfig, report))
+}
+
+var models = []struct {
+	key          string
+	schema       *jsonschema.Schema
+	modelDecoder func(interface{}, error) (transform.Transformable, error)
+}{
+	{
+		"transaction",
+		transaction.ModelSchema(),
+		transaction.DecodeEvent,
+	},
+	{
+		"span",
+		span.ModelSchema(),
+		span.DecodeSpan,
+	},
+	{
+		"metric",
+		metric.ModelSchema(),
+		metric.DecodeMetric,
+	},
+	{
+		"error",
+		er.ModelSchema(),
+		er.DecodeEvent,
+	},
+}
+
+type v2Handler struct {
+	requestDecoder decoder.ReqDecoder
+	tconfig        transform.Config
+}
+
+// handleRawModel validates and decodes a single json object into its struct form
+func (v *v2Handler) handleRawModel(rawModel map[string]interface{}) (transform.Transformable, serverResponse) {
+	for _, model := range models {
+		if entry, ok := rawModel[model.key]; ok {
+			err := validation.Validate(entry, model.schema)
+			if err != nil {
+				return nil, cannotValidateResponse(err)
+			}
+
+			tr, err := model.modelDecoder(entry, err)
+			if err != nil {
+				return tr, cannotDecodeResponse(err)
+			}
+			return tr, serverResponse{}
+		}
+	}
+	return nil, cannotValidateResponse(errors.New("did not recognize object type"))
+}
+
+// readBatch will read up to `batchSize` objects from the ndjson stream
+// it returns a slice of eventables, a serverResponse and a bool that indicates if we're at EOF.
+func (v *v2Handler) readBatch(batchSize int, reader *decoder.NDJSONStreamReader, response *streamResponse) ([]transform.Transformable, bool) {
+	var err error
+	var rawModel map[string]interface{}
+
+	eventables := []transform.Transformable{}
```

Comment:
If you are using `var eventables []transform.Transformable` instead of creating an empty array, then the formerly experienced CPU load and queuing of non-existing events disappears. 

Still think there are more issues here though, as `eof` seems to not be read properly,

Reply:
thx to @graphaelli for pointing out the cpu usage is going on in this method.